R/conditions.R

```diff
@@ -130,6 +130,59 @@ captureStackTraces <- function(expr) {
 #' @include globals.R
 .globals$deepStack <- NULL
 
+getCallStackDigest <- function(callStack, warn = FALSE) {
+  dg <- attr(callStack, "digest", exact = TRUE)
+  if (!is.null(dg)) {
+    return(dg)
+  }
+
+  if (isTRUE(warn)) {
+    rlang::warn(
+      "Call stack doesn't have a cached digest; expensively computing one now",
+      .frequency = "once",
+      .frequency_id = "deepstack-uncached-digest-warning"
+    )
+  }
+
+  rlang::hash(getCallNames(callStack))
+}
+
+saveCallStackDigest <- function(callStack) {
+  attr(callStack, "digest") <- getCallStackDigest(callStack, warn = FALSE)
+  callStack
+}
+
+# Appends a call stack to a list of call stacks, but only if it's not already
+# in the list. The list is deduplicated by digest; ideally the digests on the
+# list are cached before calling this function (you will get a warning if not).
+appendCallStackWithDedupe <- function(lst, x) {
+  tryCatch({
+    digests <- vapply(lst, getCallStackDigest, character(1), warn = TRUE)
+    xdigest <- getCallStackDigest(x, warn = FALSE)
+    stopifnot(all(nzchar(digests)))
+    stopifnot(length(xdigest) == 1)
+    stopifnot(nzchar(xdigest))
+    if (xdigest %in% digests) {
+      return(lst)
+    } else {
+      return(c(lst, list(x)))
+    }
+  }, error = function(e) {
+    # Hopefully will never hit this. I'm only putting this in to be suuuper
+    # conservative about not letting our new deep stack trace dedupe logic crash
+    # the Shiny app.
+    if (in_devmode()) {
+      printError(e, full = TRUE)
+    }
+    rlang::warn(
+      "Failed to deduplicate deep stack traces; appending anyway",
```

Comment:
This tryCatch doesn't feel overly paranoid to me, but I'm not familiar enough with this to know what risks we're running.

That said, the warning doesn't feel very actionable. If someone hits this error and it causes problems, what could they do? It'd be nice to include that in the message, e.g.

```suggestion
      c(
        "Failed to deduplicate deep stack traces; appending anyway.",
        "i" = "If this causes issues, disable deep strack traces with `options(...)`."
      ),
```

Reply:
The actionable thing to do is file an issue I guess? I had a coworker who used to put in assertions that said "Call Brian"--this is kind of like that situation 😄